File: WDWebViewBridge/WDWebViewBridge/WDWebViewBridge.js
;(function() {
  var AppBridge = {
        init:function(){
        var appBridge = {};
        appBridge.bridgeArray = new Array();
        var WDBridge = {
                init:function(){
                var bridge = {};
                bridge.time_key = new Date().getTime();
                bridge.sendMessage;
                bridge.calback;
                //发送消息到app
                bridge.sendMessageToApp = function(message){
                    var url = "wdbridgesend://" + encodeURIComponent(message)
                    window.location.href = url;
                }
                //app接收到消息后的回调
                bridge.appReceiveFinish = function(message){
                    resposeAppReceiveFinish(message);
                }
                //监听app端发来的消息
                bridge.monitorAppMessage = function(message){
                    appMonitorAppMessage(message);
                }
                //发来之后给服务端一个回调响应，表示成功
                bridge.responseToApp = function(message){
                    var url = "wdbridgeresponse://" + encodeURIComponent(message)
                    window.location.href = url;
                }
                return bridge;
            }
        }
        //给app发消息
        appBridge.sendMessageToApp = function sendMessageToApp(message,calback){
            var bridge = WDBridge.init();
            bridge.calback =  calback;
            bridge.sendMessage = message;
            appBridge.bridgeArray.push(bridge);
            //重新封装json
            var json = "{\"time_key\":"+bridge.time_key+",\"message\":\""+message+"\"}";
            //发给app
            bridge.sendMessageToApp(json);
        }
  
        //app收到消息后的回调
        appBridge.appReceiveFinish = function appReceiveFinish(message){
            var messageObj = JSON.parse(message);
            //找到对应发出消息的bridge
            for(var i = 0;i<appBridge.bridgeArray.length;i++){
                var object = appBridge.bridgeArray[i];
                if(messageObj.time_key == object.time_key){
                //调用回调方法
                object.calback(object.sendMessage);
                }
            }
        }
        appBridge.receiveAppMessage;
        //监听app的发过来的消息
        appBridge.monitorAppMessage = function monitorAppMessage(message){
            var bridge = WDBridge.init();
            var messageObj = JSON.parse(message);
            var backMessage = messageObj.sendMessage;
            //将返回过来的消息通知服务器
            bridge.responseToApp(message);
            appBridge.receiveAppMessage(backMessage);
        }
        return appBridge;
    }
  }
  window.AppBridge = AppBridge.init();
})();

File: WDWebViewBridge/WDWebViewBridge/WDWebViewBridge.m
//
//  WDWebViewBridge.m
//  WDWebViewBridge
//
//  Created by 何伟东 on 15/8/3.
//  Copyright (c) 2015年 何伟东. All rights reserved.
//

#import "WDWebViewBridge.h"
#import "WDMessage.h"

@implementation WDWebViewBridge

-(instancetype)initWithBridgeWebView:(UIWebView*)webView{
    self = [super init];
    if (self) {
        [self setBridgeWebView:webView];
        [webView setDelegate:self];
        _messageArray = [[NSMutableArray alloc] init];
    }
    return self;
}

-(void)webViewDidStartLoad:(UIWebView *)webView{
    //植入js
    NSString *bundlePath = [[NSBundle mainBundle]pathForResource:@"WDWebViewBridge" ofType:@"js"];
    NSString *javaScript = [NSString stringWithContentsOfFile:bundlePath encoding:NSUTF8StringEncoding error:nil];
    
    [webView stringByEvaluatingJavaScriptFromString:javaScript];
    if ([_delegate respondsToSelector:@selector(wdWebViewDidStartLoad:)]) {
        [_delegate wdWebViewDidStartLoad:webView];
    }
}

-(void)webViewDidFinishLoad:(UIWebView *)webView{
    if ([_delegate respondsToSelector:@selector(wdWebViewDidFinishLoad:)]) {
        [_delegate wdWebViewDidFinishLoad:webView];
    }
}

-(BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType{
    //过的URL解码
    NSString *url = [[[request URL] absoluteString] stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    //接收到js发来的消息
    if ([url hasPrefix:@"wdbridgesend://"]) {
        //js发来的消息，为任意格式的字符串
        NSString *message = [url stringByReplacingOccurrencesOfString:@"wdbridgesend://" withString:@""];
        if ([_delegate respondsToSelector:@selector(wdBridgelistener:message:)]) {
            [_delegate wdBridgelistener:self message:message];
        }
        return NO;
    }
    //app发出去消息的回调
    if ([url hasPrefix:@"wdbridgeresponse://"]) {
        //回调过来的信息
        NSString *message = [url stringByReplacingOccurrencesOfString:@"wdbridgeresponse://" withString:@""];
        NSData *jsonData = [message dataUsingEncoding:NSUTF8StringEncoding];
        NSError *err;
        NSMutableDictionary *backObject = [NSJSONSerialization JSONObjectWithData:jsonData
                                                            options:NSJSONReadingMutableContainers
                                                              error:&err];
        WDMessage *currentMessage = nil;
        for (WDMessage *message in _messageArray) {
            if ([backObject[@"time_key"] isEqualToString:message.time_key]) {
                currentMessage = message;
            }
        }
        currentMessage.sendBackBlock(backObject[@"sendMessage"]);
        [_messageArray removeObject:currentMessage];
        return NO;
    }
    
    if ([_delegate respondsToSelector:@selector(wdWebView:shouldStartLoadWithRequest:navigationType:)]) {
        [_delegate wdWebView:webView shouldStartLoadWithRequest:request navigationType:navigationType];
    }
    return YES;
}

-(void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error{
    if ([_delegate respondsToSelector:@selector(wdWebView:didFailLoadWithError:)]) {
        [_delegate wdWebView:webView didFailLoadWithError:error];
    }
}

-(void)wdBridgeSend:(NSString*)sendMessage response:(WDWebViewBridgeFinish)responseBlock{
    WDMessage *message = [[WDMessage alloc] init];
    [message setSendBackBlock:responseBlock];
    [message setContent:sendMessage];
    [_messageArray addObject:message];
    
    NSMutableDictionary *messageDictionary = [NSMutableDictionary dictionary];
    [messageDictionary setObject:message.time_key forKey:@"time_key"];
    [messageDictionary setObject:message.content forKey:@"sendMessage"];
    
    NSError *parseError = nil;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:messageDictionary options:NSJSONWritingPrettyPrinted error:&parseError];
    NSString *json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    
    NSString *javaScript = [NSString stringWithFormat:@"AppBridge.monitorAppMessage('%@')",json];
    javaScript = [javaScript stringByReplacingOccurrencesOfString:@"\n" withString:@""];
    javaScript = [javaScript stringByReplacingOccurrencesOfString:@"\r" withString:@""];
    javaScript = [javaScript stringByReplacingOccurrencesOfString:@" " withString:@""];
    [_bridgeWebView stringByEvaluatingJavaScriptFromString:javaScript];
}

-(void)resposeMessgeToJavaScript:(NSString*)responseMessage{
    NSString *message = [NSString stringWithFormat:@"AppBridge.appReceiveFinish('%@')",responseMessage];
    [_bridgeWebView stringByEvaluatingJavaScriptFromString:message];
}

@end
